domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt

```diff
@@ -0,0 +1,465 @@
+    arrangeMediaPlayer()
+
+    shadowMediaPlayer.invokeCompletionListener()
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.isSuccess()).isTrue()
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().type).isEqualTo(PlayStatus.COMPLETED)
+  }
+
+  @Test
+  fun testAudioObserver_invokeChangeDataSource_capturesPendingState() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.changeDataSource(TEST_URL2)
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.isPending()).isTrue()
+  }
+
+  @Test
+  fun testAudioObserver_invokeChangeDataSourceAfterPlay_capturesPendingState() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+    audioPlayerController.changeDataSource(TEST_URL2)
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.isPending()).isTrue()
+  }
+
+  @Test
+  fun testAudioObserver_invokePlay_capturesPlayingState() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.isSuccess()).isTrue()
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().type).isEqualTo(PlayStatus.PLAYING)
+  }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testAudioObserver_invokePlayAndWait_capturesManyPlayingState() = runBlockingTest(coroutineContext){
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+    advanceTimeBy(1000) //Wait for next schedule update call
+    shadowMediaPlayer.invokeCompletionListener()
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.allValues.size).isEqualTo(4)
+    assertThat(audioPlayerResultCaptor.allValues[0].getOrThrow().type).isEqualTo(PlayStatus.PREPARED)
+    assertThat(audioPlayerResultCaptor.allValues[1].getOrThrow().type).isEqualTo(PlayStatus.PLAYING)
+    assertThat(audioPlayerResultCaptor.allValues[2].getOrThrow().type).isEqualTo(PlayStatus.PLAYING)
+    assertThat(audioPlayerResultCaptor.allValues[3].getOrThrow().type).isEqualTo(PlayStatus.COMPLETED)
+    assertThat(audioPlayerResultCaptor.value.isSuccess()).isTrue()
+  }
+
+  @Test
+  fun testAudioObserver_invokePause_capturesPausedState() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+    audioPlayerController.pause()
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.isSuccess()).isTrue()
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().type).isEqualTo(PlayStatus.PAUSED)
+  }
+
+  @Test
+  fun testAudioObserver_invokePrepared_capturesCorrectProgress() {
```

Comment:
Can you also add a test to verify that progress correctly changes over time after waiting for the update job to run? Or can the shadow not do this?